File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null) return root;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        while(!queue.isEmpty()) {
            int size = queue.size();
            while(size-- > 0) {
                Node temp = queue.poll();
                if(size != 0) temp.next = queue.peek();
                if(temp.left != null) queue.offer(temp.left);
                if(temp.right != null) queue.offer(temp.right);
            }
        }
        return root;
    }
}

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root: return root
        q = deque()
        q.append(root)
        while q:
            size = len(q)
            while size:
                size -= 1
                temp = q.popleft()
                if size: temp.next = q[0]
                if temp.left: q.append(temp.left)
                if temp.right: q.append(temp.right)
        return root